src/cms.c

```diff
@@ -0,0 +1,96 @@
+#include <stdlib.h>         // malloc
+#include <string.h>         // strlen
+#include <stdio.h>          // printf
+#include <math.h>           // q, ceil
+#include <assert.h>         // assert
+
+#include "cms.h"
+#include "contrib/xxhash.h"
+
+
+CMSketch *NewCMSketch(size_t width, size_t depth) {
+    assert(width > 0);
+    assert(depth > 0);
+
+    CMSketch *cms = CMS_CALLOC(1, sizeof(CMSketch));
+
+    cms->width = width;
+    cms->depth = depth;
+    cms->counter = 0;
+    cms->array = CMS_CALLOC(width * depth, sizeof(size_t));
+
+    return cms; 
+}
+
+void CMS_Destroy(CMSketch *cms) {
+    assert(cms);
+
+    CMS_FREE(cms->array);
+    cms->array = NULL;
+
+    CMS_FREE(cms);
+}
+
+void CMS_IncrBy(CMSketch *cms, const char *item, size_t value) {
```

Comment:
Accept a buffer and a length, rather than assume a NUL-terminated string. e.g. `const char *item, size_t itemlen`, and then pass `itemlen` to the hash function.